Merge profiles onto the Meteor.User object

ebroder opened the pull request:
We historically avoided using the profile field on Meteor.Users due to
its relatively lax default permissioning scheme. But those defaults are
not difficult to change, and having user information spread across two
different collections makes a lot of query patterns difficult (since
they effectively require joins).

So instead, migrate the old profiles model onto the Meteor.User object.
This is intended to be a fairly 1-for-1 migration in order to make
review easier, so it does not do things like add additional constraints
to data visibility relative to what we have historically done; that's
reserved as a future task.

Fixes #628.

ebroder commented on imports/server/users.ts:
```diff
@@ -29,6 +29,31 @@ Meteor.publish('huntMembers', function (huntId: string) {
   return MeteorUsers.find({ hunts: huntId }, { fields: { hunts: 1 } });
 });
 
+Meteor.publish('displayNames', function () {
+  if (!this.userId) {
+    return [];
+  }
+
+  return MeteorUsers.find({}, { fields: { 'profile.displayName': 1 } });
+});
+
+Meteor.publish('avatars', function () {
+  if (!this.userId) {
+    return [];
+  }
+
+  return MeteorUsers.find({}, { fields: { 'profile.discordAccount': 1 } });
+});
+
+Meteor.publish('profiles', function () {
+  if (!this.userId) {
+    return [];
+  }
+
+  // For now, all user profiles are public, including email address
+  return MeteorUsers.find({}, { fields: { 'emails.address': 1, profile: 1 } });
+});
+
```
Given the sensitivity around the users model, I decided that life would probably be easier if we had a more more-limited publishes rather than the singular "pass your own query" funtimes from the old `mongo.profiles` publish. (I'm not yet proposing an official style change yet, but in general I'm a little over the "pass your own query" style of publishes.). Long-term, my plan is to make these take a `huntId` parameter and then do some :sparkles:query and projection magic:sparkles:, but reserving that for future work. For now, this just preserves exactly the same functionality as we had before.

zarvox commented on imports/client/components/AllProfileListPage.tsx:
```diff
@@ -1,22 +1,22 @@
-  const profiles = useTracker(() => {
-    return loading ? [] : Profiles.find({}, { sort: { displayName: 1 } }).fetch();
+  const users = useTracker(() => {
+    return loading ? [] : Meteor.users.find({}, { sort: { 'profile.displayName': 1 } }).fetch();
```
Any reason this uses `Meteor.users` directly instead of `MeteorUsers`?

zarvox commented on imports/client/components/App.tsx:
```diff
@@ -156,16 +155,8 @@ const ErrorFallback = ({ error, resetErrorBoundary }: FallbackProps) => {
 
 const AppNavbar = () => {
   const userId = useTracker(() => Meteor.userId()!, []);
-  const profileLoading = useSubscribe('mongo.profiles', { _id: userId });
```
Oh nice this saves a roundtrip!

zarvox commented on imports/client/components/CallSection.tsx:
```diff
@@ -175,10 +174,10 @@ const ProducerBox = ({
     return {
-      initial: profile ? profile.displayName.slice(0, 1) : 'U', // get it?  it's you
-      discordAvatarUrl: getAvatarCdnUrl(profile?.discordAccount),
+      initial: user.profile?.displayName ?? 'U', // get it?  it's you
```
```suggestion
      initial: user.profile?.displayName?.slice(0, 1) ?? 'U', // get it?  it's you
```

zarvox commented on imports/client/components/ChatPeople.tsx:
```diff
@@ -217,24 +216,23 @@ const ChatPeople = ({
-      const user = p.createdBy;
-      const profile = Profiles.findOne(user);
-      if (!profile || !profile.displayName) {
+      const user = Meteor.users.findOne(p.createdBy);
```
`MeteorUsers` instead of `Meteor.users`?

zarvox commented on imports/server/profile.ts:
```diff
@@ -22,23 +21,16 @@ Meteor.methods({
-
     Ansible.log('Updating profile for user', { user: this.userId });
-    Profiles.update({
+    Meteor.users.update({
```
Got some more uses of `Meteor.users` in this file -- is this just inconsistency, or is there a set of situations where we need to use `Meteor.users` directly instead of `MeteorUsers`?

zarvox commented on imports/server/migrations/38-consolidate-profiles.ts:
```diff
@@ -0,0 +1,28 @@
+    });
+
+    Profiles.find({}).forEach((profile) => {
+      Meteor.users.update(profile._id, { $set: { profile } });
```
If I'm not mistaken: this sets a field `_id` on the nested profile object?  Maybe pluck just the fields we actually expect here instead?

zarvox commented:
A couple higher-level things:

It's not clear to me if you're still intending to avoid using `Meteor.users` directly (so we're guaranteed that we pull in the deny rule and simpl-schema attachment early).  I think we likely still want to; if so, you should go back through this and do the bulk replace.

I'm somewhat skeptical the `defaultValue` on `displayName` from `imports/lib/schemas/user.ts` is actually helping much -- any time that we'd not have a name for a user, we should indicate the absence instead of just using the empty string.  It'd be good to test the situation where a user gets into a situation where they have no display name (missed their initial invitation email and then used password reset to recover their account but as a result never did the profile onboarding) and see what result that produces.  We should also make sure that `ProfileMissingMessage` in `NotificationCenter` still guides the user to populate a real profile (and perhaps even a non-empty `displayName` if the defaultValue here is automatically populating it with empty-string).

ebroder commented:
Alright - I went ahead and switched everything to use `MeteorUsers`, and I manually copied the fields I wanted in the migration.

For `displayName`, I dropped the default value. I think the schema now expresses that "you may not have a profile at all, but if you have a profile, it must have a `displayName` field", which seems like a fine semantic to have. I also verified that if I unset (`db.users.update({_id: '...'}, {$unset: {profile: 1}})`) the profile field or set it to an empty object (`db.uses.update({_id: '...'}, {$set: {profile: {}}})`, which should theoretically be impossible), I get the missing profile notification.
